src/Juvix/Compiler/Internal/Extra/Clonable.hs

```diff
@@ -0,0 +1,260 @@
+  freshNameIds Let {..} = do
+    underClonableBindersNonEmpty _letClauses $ \clauses' -> do
+      e' <- freshNameIds _letExpression
+      return
+        Let
+          { _letClauses = clauses',
+            _letExpression = e'
+          }
+
+instance Clonable SimpleBinder where
+  freshNameIds SimpleBinder {..} = do
+    ty' <- freshNameIds _sbinderType
+    return
+      SimpleBinder
+        { _sbinderType = ty',
+          ..
+        }
+
+instance Clonable SimpleLambda where
+  freshNameIds SimpleLambda {..} =
+    underBinder _slambdaBinder $ \bi -> do
+      bi' <- freshNameIds bi
+      body' <- freshNameIds _slambdaBody
+      return
+        SimpleLambda
+          { _slambdaBinder = bi',
+            _slambdaBody = body'
+          }
+
+instance Clonable LambdaClause where
+  freshNameIds LambdaClause {..} =
+    underBindersNonEmpty _lambdaPatterns $ \ps' -> do
+      body' <- freshNameIds _lambdaBody
+      return
+        LambdaClause
+          { _lambdaPatterns = ps',
+            _lambdaBody = body'
+          }
+
+instance Clonable Lambda where
+  freshNameIds Lambda {..} = do
+    ty' <- freshNameIds _lambdaType
+    clauses' <- freshNameIds _lambdaClauses
+    return
+      Lambda
+        { _lambdaType = ty',
+          _lambdaClauses = clauses'
+        }
+
+instance Clonable Expression where
+  freshNameIds :: (Members '[State HolesState, Reader FreshBindersContext, NameIdGen] r) => Expression -> Sem r Expression
+  freshNameIds = \case
+    ExpressionIden i -> ExpressionIden <$> freshNameIds i
+    ExpressionApplication a -> ExpressionApplication <$> freshNameIds a
+    ExpressionLiteral a -> ExpressionLiteral <$> freshNameIds a
+    ExpressionHole a -> ExpressionHole <$> freshNameIds a
+    ExpressionUniverse a -> ExpressionUniverse <$> freshNameIds a
+    ExpressionCase a -> ExpressionCase <$> freshNameIds a
+    ExpressionFunction f -> ExpressionFunction <$> freshNameIds f
+    ExpressionInstanceHole h -> ExpressionInstanceHole <$> freshNameIds h
+    ExpressionLet l -> ExpressionLet <$> freshNameIds l
+    ExpressionSimpleLambda l -> ExpressionSimpleLambda <$> freshNameIds l
+    ExpressionLambda l -> ExpressionLambda <$> freshNameIds l
+
+instance Clonable ArgInfo where
+  freshNameIds ArgInfo {..} = do
+    sig' <- mapM freshNameIds _argInfoDefault
+    return
+      ArgInfo
+        { _argInfoDefault = sig',
+          _argInfoName
+        }
+
+instance Clonable FunctionDef where
+  freshNameIds :: (Members '[State HolesState, Reader FreshBindersContext, NameIdGen] r) => FunctionDef -> Sem r FunctionDef
+  freshNameIds fun@FunctionDef {..} = do
+    ty' <- freshNameIds _funDefType
+    underBinder fun $ \fun' -> do
+      body' <- freshNameIds _funDefBody
+      defaultSig' <- freshNameIds _funDefArgsInfo
+      return
+        FunctionDef
+          { _funDefName = fun' ^. funDefName,
+            _funDefType = ty',
+            _funDefBody = body',
+            _funDefArgsInfo = defaultSig',
+            ..
```

Comment:
I think the `..` is dangerous here. One needs to remember that all `Expression` fields need to be refreshed here. It's easy to forget when one adds a new field, and with `..` the compiler won't give an error but silently use the original non-refreshed field.